File: C++/05_함수/Problem_02.cpp
/* 05단계 [ 함수 ] */
// 제목 : 셀프 넘버
// 문제 번호 : 4673

#include <iostream>
using namespace std;

#define array_size 10001
bool self_num[array_size];

int NSN(int x) // Not a Self Number : 생성자를 반환
{
    int sum = x;
    while (x)
    {
        sum += x % 10;
        x /= 10;
    }
    return sum;
}

int main(void)
{
    for (int i = 1; i < array_size; i++) // 모든숫자를 'self number'로 저장
        self_num[i] = true;
    for (int i = 1; i < array_size; i++) // 1~N 까지의 생성자를 'self number'에서 제외
        if (NSN(i) < array_size)
            self_num[NSN(i)] = false;
    for (int i = 1; i < array_size; i++) // 'self number'출력
        if (self_num[i])
            cout << i << endl;
    return 0;
}

File: Python/05_함수/Problem_02.py
def NSN(x):
    sum = x
    while x != 0:
        sum += x % 10
        x //= 10
    return sum


self_num = [False]
for i in range(10000):
    self_num.append(True)

for i in range(1, 10001):
    if NSN(i) <= 10000:
        self_num[NSN(i)] = False

for i in range(1, 10001):
    if self_num[i]:
        print(i)
